Fixes #31596 - Timeseries Area chart in React

ezr-ondrej opened the pull request:
Implements area chart, previously implemented by `flot_chart` in React.

Main user should be: https://github.com/theforeman/foreman_statistics/pull/25

ezr-ondrej committed:
Fixes #31596 - Timeseries Area chart in React

Implements area timeseries chart, previously implemented by `flot_chart` in React.

amirfefer commented on webpack/assets/javascripts/services/charts/AreaChartService.js:
```diff
@@ -0,0 +1,52 @@
+  if (config === 'timeseries' && xAxisDataLabel) {
+    chartConfig.data.x = xAxisDataLabel;
+  } else if (config === 'timeseries' && !xAxisDataLabel) {
+    throw new Error('xAxisDataLabel is missing for timeseries area graph');
```
wouldn't throwing a warning be safer?

amirfefer commented on webpack/assets/javascripts/services/charts/AreaChartService.js:
```diff
@@ -0,0 +1,52 @@
+    id,
+  });
+
+  if (config === 'timeseries' && xAxisDataLabel) {
+    chartConfig.data.x = xAxisDataLabel;
+  } else if (config === 'timeseries' && !xAxisDataLabel) {
```
`xAxisDataLabel` has a default value, it can't be undefined, no?

ezr-ondrej replied:
we could get the undef from the caller, but it would be quite dumb from the caller, so we probably don't need to worry about that :thinking:

amirfefer commented on webpack/assets/javascripts/services/charts/AreaChartService.js:
```diff
@@ -0,0 +1,52 @@
+    chartConfig.data.colors = {};
+    chartConfig.data.columns[0] = [xAxisDataLabel].concat(
+      data[0].slice(1).map(epochSecs => new Date(epochSecs * 1000))
+    );
```
how about make it a bit simpler? for increasing readability
Something like
```js
const datesInit = data[0].slice(1).map(epochSecs => new Date(epochSecs * 1000))
chartConfig.data.columns[0] = [xAxisDataLabel].concat(formatedDattes)
```

amirfefer commented on webpack/assets/javascripts/services/charts/ChartService.consts.js:
```diff
@@ -35,7 +35,7 @@ export const chartConfig = {
   },
   color: {
-    pattern: ['#0088ce', '#ec7a08', '#3f9c35', '#005c66', 'f9d67a', '#703fec'],
+    pattern: ['#0088ce', '#ec7a08', '#3f9c35', '#005c66', '#f9d67a', '#703fec'],
```
good catch :)

amirfefer reviewed (changes requested):
Thanks @ezr-ondre!j 👍 

I left a few comments, mostly nits.
Could you please add a storybook and some tests for this chart? there is no other use for this, and a story would be a great way to verify it.

ezr-ondrej commented:
I've rebased, added test for the service and chart to the storybook.
The component just takes props, passes them to the service and push the result to C3, so I guess we don't need to test that :thinking:

ezr-ondrej commented:
[test upgrade]

ezr-ondrej commented:
[test katello]

ezr-ondrej commented:
Adding tag as demo worthy not to forget to demo the statistics plugin that uses this.

amirfefer commented:
[test foreman]

amirfefer reviewed (approved):
Thanks, @ezr-ondrej, tested on storybook and works well 👍 
Overall I think these PF3 charts are outdated and the `ChartService` helper adds a redundant complexity.  I'll raise this issue and the alternative PF4 charts (finally, a maintained [library](https://formidable.com/open-source/victory/) than c3) to the next UI group meeting.
